fix(timeseries): prevent automatic coarsening of monthly ticks when there’s capacity (closes #60475)

lwensveen opened the pull request:
Closes [https://github.com/metabase/metabase/issues/60475](https://github.com/metabase/metabase/issues/60475)

## Description

On time series charts with a 1-month bucket over roughly a year, ECharts was automatically coarsening the tick interval (e.g., showing every 2 months / 6 labels) even when there was enough horizontal space for all 12 monthly labels. This change adds a heuristic in `getTicksOptions` to detect the exact 1-month interval with sufficient capacity (>=12) and pins both `minInterval` and `maxInterval` to the one-month duration so that ECharts cannot coarsen it. When space allows, all 12 months are shown; when space is constrained it still degrades to coarser intervals.

## How to verify

1. Run the frontend locally and open a new question on the Sample Dataset (e.g., Orders).
2. Bucket the x-axis by "Created At: Month" covering a 12-month span (e.g., Jan–Dec).
3. View it as a bar/time-series chart.
4. With the chart wide enough, verify that all 12 month labels (Jan through Dec) appear on the x-axis.
5. Shrink the chart width and confirm that it gracefully degrades (e.g., drops to fewer labels like 6) when there isn’t space.
6. Verify other time ranges (shorter than 12 months) still behave reasonably and don’t regress.

## Demo

*Before*: Monthly buckets were coarsened to every 2 months even when the chart was wide enough.
<img width="1679" height="964" alt="Screenshot 2568-07-31 at 12 34 37" src="https://github.com/user-attachments/assets/10e9dfdf-a142-4f5a-ab1d-811865e5d032" />

*After*: All 12 month ticks are shown when capacity allows.
<img width="1680" height="949" alt="Screenshot 2568-07-31 at 16 27 06" src="https://github.com/user-attachments/assets/2c6a1e92-5e77-43f6-abe5-539ed2997be9" />
<img width="1679" height="960" alt="Screenshot 2568-07-31 at 16 27 18" src="https://github.com/user-attachments/assets/dcd60941-6f6e-4604-8fd2-c6a05b41fd7f" />


## Checklist

* [ ] Tests have been added/updated to cover changes in this PR
  *Note: No existing test coverage for this code path; none were added in this change because the file and environment are large to validate locally. Will add if requested by reviewer.*

lwensveen committed:
fix(timeseries): pin 1-month ticks when capacity allows (#60475)

deniskaber reviewed (approved):
Added some minor comments, otherwise looks good!

lwensveen committed:
Update frontend/src/metabase/visualizations/echarts/cartesian/model/util.ts

Co-authored-by: Denis Berezin <deniskaber@mail.ru>

lwensveen committed:
Update frontend/src/metabase/visualizations/echarts/cartesian/option/ticks.ts

Co-authored-by: Denis Berezin <deniskaber@mail.ru>

deniskaber reviewed (changes requested):
I ran CI for your PR in https://github.com/metabase/metabase/pull/61986 , looks like there are some issues with linter and unit tests.  So, you might have to adjust logic you added.

lwensveen commented:
> I ran CI for your PR in #61986 , looks like there are some issues with linter and unit tests. So, you might have to adjust logic you added.

Sorry, I was busy with some job applications the last few days. As soon as I got the backend drivers downloaded and built, which takes some time, I'll fix the tests.

lwensveen committed:
echarts(timeseries): treat annual series as year(1) and prefer year over 12 months
- Add helpers: coerceYear (12 months -> 1 year) and coerceAnnualFromMonthly (same M/D across years -> year)
- Apply coercion in computeTimeseriesDataInterval and computeTimeseriesTicksInterval
- Reorder TIMESERIES_INTERVALS so year(2) is checked right after year(1) before multi-month entries
- Keep logic otherwise unchanged

lwensveen committed:
echarts(ticks): only pin monthly ticks when computed interval is month(1)
- Prevents accidental monthly grid on annual series (min/max interval clamp now gated by computedInterval)

lwensveen committed:
echarts(model): tidy import of expectedTickCount/maxTicksForChartWidth (no behavior change)

lwensveen commented:
Okay, I made some changes, and it seems to be fixed, but I'm not 100% sure. Since after running these E2E tests my laptop nearly explodes, I'm temporarily working on a 8GB M1 since I'm on vacation. So I could only verify what you see below. If there's any issues remaining, I'm sorry. But I think the changes I've made should logically fix the issues we saw before.

<img width="1047" height="291" alt="Screenshot 2568-08-13 at 14 58 43" src="https://github.com/user-attachments/assets/055534ad-fa8d-49eb-b3fb-7c24115728db" />
